File: exercise/kotlin/day24/src/main/kotlin/domain/Toy.kt
package domain

import arrow.core.Either
import arrow.core.left
import arrow.core.right
import domain.core.Error
import domain.core.EventSourcedAggregate
import java.time.LocalDateTime
import java.util.*

class Toy private constructor(
    timeProvider: () -> LocalDateTime,
    private var name: String?,
    private var stock: StockUnit?
) : EventSourcedAggregate(timeProvider) {
    fun name(): String = name!!

    init {
        raiseEvent(ToyCreatedEvent(UUID.randomUUID(), timeProvider(), name!!, stock!!))
    }

    companion object {
        fun create(timeProvider: () -> LocalDateTime, name: String, stock: Int): Either<Error, Toy> =
            StockUnit.from(stock)
                .map { stockUnit ->
                    Toy(timeProvider, name, stockUnit)
                }
    }

    private fun apply(event: ToyCreatedEvent) {
        id = event.id
        name = event.name
        stock = event.stock
    }

    fun reduceStock(): Either<Error, Toy> {
        if (stock?.isSupplied() == false) return Error("No more $name in stock").left()
        raiseEvent(StockReducedEvent(id, time(), name!!, stock!!.decrease()))

        return right()
    }

    private fun apply(event: StockReducedEvent) {
        stock = event.newStock
    }

    override fun registerRoutes() {
        registerEventRoute(ToyCreatedEvent::class.java, this::apply)
        registerEventRoute(StockReducedEvent::class.java, this::apply)
    }
}

File: exercise/ts/day24/src/domain/Toy.ts
import {Either, left, right} from 'fp-ts/lib/Either';
import {EventSourcedAggregate} from "./core/EventSourcedAggregate";
import {StockUnit} from "./StockUnit";
import {ToyCreatedEvent} from "./ToyCreatedEvent";
import {StockReducedEvent} from "./StockReducedEvent";
import {pipe} from 'fp-ts/function';
import * as E from "fp-ts/Either";
import {DomainError} from "./core/DomainError";

export class Toy extends EventSourcedAggregate {
    name?: string;
    private stock: StockUnit;

    private constructor(timeProvider: () => Date, name: string, stock: StockUnit) {
        super(timeProvider);
        this.raiseEvent(new ToyCreatedEvent(this.id, this.time(), name, stock));
    }

    static create(timeProvider: () => Date, name: string, stock: number): Either<DomainError, Toy> {
        return pipe(
            StockUnit.from(stock),
            E.map(s => new Toy(timeProvider, name, s))
        );
    }

    private applyToyCreated(toy: Toy, event: ToyCreatedEvent): void {
        toy.id = event.id;
        toy.name = event.name;
        toy.stock = event.stock;
    }

    reduceStock(): Either<DomainError, Toy> {
        if (!this.stock.isSupplied()) return left(new DomainError(`No more ${this.name} in stock`));
        this.raiseEvent(new StockReducedEvent(this.id, this.time(), this.name!, this.stock.decrease()));

        return right(this);
    }

    private applyStockReduced(toy: Toy, event: StockReducedEvent): void {
        toy.stock = event.newStock;
    }

    protected registerRoutes(): void {
        this.registerEventRoute(ToyCreatedEvent, evt => this.applyToyCreated(this, evt));
        this.registerEventRoute(StockReducedEvent, evt => this.applyStockReduced(this, evt));
    }
}